From af0084b691adc789a24d6d1a6184eeca32e4948f Mon Sep 17 00:00:00 2001
From: Yosuke Matsumura <yosukematsumura@gmail.com>
Date: Tue, 8 Apr 2025 08:28:41 -0700
Subject: [PATCH] Improvements to the reliability of auto updates

Adding a wait to refreshUnattended to allow for modules to perform
update checks

Change refreshUnattended to use lastUnattendedTrigger instead of
lastNotificationTime as basis of determining whether to run.

BUG: 447245
---
 notifier/DiscoverNotifier.cpp | 38 +++++++++++++++++++++++++++--------
 notifier/DiscoverNotifier.h   |  2 ++
 2 files changed, 32 insertions(+), 8 deletions(-)

diff --git a/notifier/DiscoverNotifier.cpp b/notifier/DiscoverNotifier.cpp
index 407614db9..6f33028ee 100644
--- a/notifier/DiscoverNotifier.cpp
+++ b/notifier/DiscoverNotifier.cpp
@@ -135,24 +135,33 @@ void DiscoverNotifier::showDiscoverUpdates(const QString &xdgActivationToken)
     }
 }
 
-bool DiscoverNotifier::notifyAboutUpdates() const
+bool DiscoverNotifier::checkTriggerTimes(const QDateTime &lastTriggerTime) const
 {
     if (state() != NormalUpdates && state() != SecurityUpdates) {
         // it's not very helpful to notify that everything is in order
-        qCDebug(NOTIFIER) << "Not notifying about updates, state is" << state();
+        qCDebug(NOTIFIER) << "Not triggering, state is" << state();
         return false;
     }
 
     if (m_settings->requiredNotificationInterval() < 0) {
-        qCDebug(NOTIFIER) << "Not notifying about updates, requiredNotificationInterval is" << m_settings->requiredNotificationInterval();
+        qCDebug(NOTIFIER) << "Not triggering, requiredNotificationInterval is" << m_settings->requiredNotificationInterval();
         return false;
     }
 
     // To configure to a random value, execute:
     // kwriteconfig5 --file PlasmaDiscoverUpdates --group Global --key RequiredNotificationInterval 3600
-    const QDateTime earliestNextNotificationTime = m_settings->lastNotificationTime().addSecs(m_settings->requiredNotificationInterval());
-    if (earliestNextNotificationTime.isValid() && earliestNextNotificationTime > QDateTime::currentDateTimeUtc()) {
-        qCDebug(NOTIFIER) << "Not notifying about updates, earliestNextNotificationTime is" << earliestNextNotificationTime;
+    const QDateTime earliestNextTriggerTime = lastTriggerTime.addSecs(m_settings->requiredNotificationInterval());
+    if (earliestNextTriggerTime.isValid() && earliestNextTriggerTime > QDateTime::currentDateTimeUtc()) {
+        qCDebug(NOTIFIER) << "Not triggering, earliestNextTriggerTime is" << earliestNextTriggerTime;
+        return false;
+    }
+
+    return true;
+}
+
+bool DiscoverNotifier::notifyAboutUpdates() const
+{
+    if (!checkTriggerTimes(m_settings->lastNotificationTime())) {
         return false;
     }
 
@@ -166,6 +175,19 @@ bool DiscoverNotifier::notifyAboutUpdates() const
     return true;
 }
 
+bool DiscoverNotifier::proceedUnattended() const
+{
+    if (!checkTriggerTimes(m_settings->lastUnattendedTrigger())) {
+        return false;
+    }
+
+    if (QDBusConnection::sessionBus().interface()->isServiceRegistered(QStringLiteral("org.kde.discover"))) {
+        qCDebug(NOTIFIER) << "Not proceeding with unattended update, discover is running";
+        return false;
+    }
+    return true;
+}
+
 void DiscoverNotifier::showUpdatesNotification()
 {
     if (m_updatesAvailableNotification) {
@@ -269,7 +291,7 @@ void DiscoverNotifier::refreshUnattended()
 {
     m_settings->read();
 
-    if (!notifyAboutUpdates()) {
+    if (!proceedUnattended()) {
         qCDebug(NOTIFIER) << "refreshUnattended: not notifying about updates";
         return;
     }
@@ -349,7 +371,7 @@ void DiscoverNotifier::recheckSystemUpdateNeeded()
     for (BackendNotifierModule *module : std::as_const(m_backends))
         module->recheckSystemUpdateNeeded();
 
-    refreshUnattended();
+    QTimer::singleShot(20000, this, &DiscoverNotifier::refreshUnattended);
 }
 
 QStringList DiscoverNotifier::loadedModules() const
diff --git a/notifier/DiscoverNotifier.h b/notifier/DiscoverNotifier.h
index 7e63faefe..59fd8de51 100644
--- a/notifier/DiscoverNotifier.h
+++ b/notifier/DiscoverNotifier.h
@@ -94,7 +94,9 @@ private:
     void updateStatusNotifier();
     void refreshUnattended();
 
+    bool checkTriggerTimes(const QDateTime &lastTriggerTime) const;
     bool notifyAboutUpdates() const;
+    bool proceedUnattended() const;
 
     QList<BackendNotifierModule *> m_backends;
     QTimer m_timer;
-- 
GitLab

